[WGSL] Textures and samplers shouldn't be references

tadeuzagallo opened the pull request:
#### 889edf206f1a3946daf6664d6b2e5870c5c9cc28
<pre>
[WGSL] Textures and samplers shouldn&apos;t be references
<a href="https://bugs.webkit.org/show_bug.cgi?id=254425">https://bugs.webkit.org/show_bug.cgi?id=254425</a>
rdar://107188253

Reviewed by Myles C. Maxfield.

When converting globals to argument buffers, most values become references, but
that doesn&apos;t apply to textures and samplers. Add logic to check the type of the
global instead of unconditionally converting it to a reference.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertStructs):

Canonical link: <a href="https://commits.webkit.org/262200@main">https://commits.webkit.org/262200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c248a69edb19a73d29c923362162af54b78e9c6f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/500 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/541 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/544 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/449 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/602 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/689 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/508 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/558 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/539 "3 api tests failed or timed out") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/476 "1 flakes 7 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/517 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/444 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/495 "9 failures") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/474 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/494 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/489 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/207 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/482 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -236,10 +236,23 @@ void RewriteGlobalVariables::insertStructs()
+            }
+
+            AST::TypeName::Ref memberType = *global.declaration->maybeTypeName();
+            if (shouldBeReference)
+                memberType = adoptRef(*new AST::ReferenceTypeName(span, WTFMove(memberType)));
```
I think this is the right design, but how confident are we that this actually works on all metal devices that we support? The output program would be something like

```
kernel void foo(texture2d<float> foo) {
    ...
    texture2d<float> bar = foo;
    ...
}
```

I guess I should try to do the research to make sure this actually works everywhere.

litherum replied:
Looks like yes this is supposed to work everywhere.

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Textures and samplers shouldn't be references
https://bugs.webkit.org/show_bug.cgi?id=254425
rdar://107188253

Reviewed by Myles C. Maxfield.

When converting globals to argument buffers, most values become references, but
that doesn't apply to textures and samplers. Add logic to check the type of the
global instead of unconditionally converting it to a reference.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertStructs):

Canonical link: https://commits.webkit.org/262200@main
